src/sources/http.rs

```diff
@@ -0,0 +1,358 @@
+use crate::{
+    event::{self, flatten::flatten, Event},
+    sources::util::{ErrorMessage, HttpSource},
+    topology::config::{DataType, GlobalOptions, SourceConfig},
+};
+use bytes::Buf;
+use chrono::Utc;
+use futures::sync::mpsc;
+use serde::{Deserialize, Serialize};
+use serde_json::Value as JsonValue;
+use std::{
+    io::{BufRead, BufReader, Read},
+    net::SocketAddr,
+};
+use warp::http::{HeaderMap, HeaderValue, StatusCode};
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+pub struct SimpleHttpConfig {
+    address: SocketAddr,
+    #[serde(default)]
+    encoding: Encoding,
+    #[serde(default)]
+    headers: Vec<String>,
+}
+
+#[derive(Clone)]
+struct SimpleHttpSource {
+    encoding: Encoding,
+    headers: Vec<String>,
+}
+
+#[derive(Deserialize, Serialize, Debug, Eq, PartialEq, Clone, Derivative, Copy)]
+#[serde(rename_all = "snake_case")]
+#[derivative(Default)]
+pub enum Encoding {
+    #[derivative(Default)]
+    Text,
+    Ndjson,
+    Json,
+}
+
+impl HttpSource for SimpleHttpSource {
+    fn build_event(
+        &self,
+        body: impl Buf,
+        header_map: HeaderMap,
+    ) -> Result<Vec<Event>, ErrorMessage> {
+        decode_body(body, self.encoding)
+            .map(|events| add_headers(events, &self.headers, header_map))
+    }
+}
+
+#[typetag::serde(name = "http")]
+impl SourceConfig for SimpleHttpConfig {
+    fn build(
+        &self,
+        _: &str,
+        _: &GlobalOptions,
+        out: mpsc::Sender<Event>,
+    ) -> crate::Result<super::Source> {
+        let source = SimpleHttpSource {
+            encoding: self.encoding,
+            headers: self.headers.clone(),
+        };
+        source.run(self.address, "".to_string(), out)
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn source_type(&self) -> &'static str {
+        "http"
+    }
+}
+
+fn add_headers(
+    mut events: Vec<Event>,
+    headers_config: &[String],
+    headers: HeaderMap,
+) -> Vec<Event> {
+    for header_name in headers_config {
+        let value = headers
+            .get(header_name)
+            .map(HeaderValue::as_bytes)
+            .unwrap_or_default();
+        for event in events.iter_mut() {
+            event.as_mut_log().insert(header_name as &str, value);
+        }
+    }
+
+    events
+}
+
+fn body_to_lines(body: impl Buf) -> impl Iterator<Item = String> {
+    BufReader::new(body.reader())
+        .lines()
+        .filter_map(|res| {
+            res.map_err(|error| error!(message = "Error reading request body", ?error))
+                .ok()
+        })
+        .filter(|s| !s.is_empty())
+}
+
+fn decode_body(body: impl Buf, enc: Encoding) -> Result<Vec<Event>, ErrorMessage> {
```

Comment:
So I think it may be a bit better here to do something like:

```rust

let body = buf.collect::<BytesMut>();
```

then for the encodings that need them in lines format we can take advantage of https://github.com/timberio/vector/blob/master/lib/codec/src/lib.rs#L18

then for the `json` encoding we can do something like:

```rust
serde_json::from_slice(&body[..])
```

This would avoid some additional copying

Reply:
ok sounds good

Reply:
new commit is mostly for this. lots of changes to the body_to_lines function, which is now not the prettiest... let me know what you think of it, if there are better ways to do what I'm doing.

one issue I had is that the codec won't give the last line if it doesn't end in a newline, so I have a special case for that.

Reply:
Our codec that we wrote should handle that, you may want to call decode_eof though instead which should handle the last line.

Reply:
oh perfect! Thanks :)